File: src/main/kotlin/easy/_88_merge_sorted_array.kt
package easy

object _88_merge_sorted_array {
    fun merge(nums1: IntArray, m: Int, nums2: IntArray, n: Int) {
        if (nums1.isEmpty() || nums2.isEmpty()) return
        var a1 = m - 1
        var a2 = n - 1

        for (i in nums1.lastIndex downTo 0) {
            if (a1 >= 0 && (a2 < 0 || nums1[a1] > nums2[a2])) {
                nums1[i] = nums1[a1--]
            } else {
                nums1[i] = nums2[a2--]
            }
        }

        println(nums1.toList())
    }
}

fun main() {
    with(_88_merge_sorted_array) {
        merge(
            intArrayOf(1, 2, 3, 0, 0, 0),
            3,
            intArrayOf(2, 5, 6),
            3
        )
    }
}

File: python/leetcode-150/88. Merge Sorted Array.py
from typing import List


class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        fill_idx = len(nums1) - 1
        l = m - 1
        r = n - 1
        while fill_idx >= 0:
            if l >= 0 and r >= 0:
                if nums1[l] >= nums2[r]:
                    nums1[fill_idx] = nums1[l]
                    l -= 1
                else:
                    nums1[fill_idx] = nums2[r]
                    r -= 1
            elif l < 0:
                nums1[fill_idx] = nums2[r]
                r -= 1
            elif r <0:
                nums1[fill_idx] = nums1[l]
                l -= 1
            fill_idx -= 1


if __name__ == "__main__":
    s = Solution()
    nums1 = [1, 2, 3, 0, 0, 0]
    nums2 = [2, 5, 6]
    s.merge(nums1, 3, nums2, 3)
    print(nums1)
    print(nums2)

    nums1 = [0]
    nums2 = [1]
    s.merge(nums1, 0, nums2, 1)
    print(nums1)
    print(nums2)
